### ISSUE TITLE
proposed page: Creating a front-end theme

### ISSUE THREAD

**Issue by wion** [textpattern/textpattern.github.io#128]
Definitely a document that needs written...

> I’m so far finding the answers, but only after looking all over the place. It’d be nice if there was one big “This is basically what you need,” type of document/page.
> – [craigbass76](https://forum.textpattern.com/viewtopic.php?pid=316470#p316470), the tech writer

## What will be the scope and structure of document?

* What is needed to start?
* What is the tutorial completion point?
* What are the essential sections of the document body?

### Existing resources to review or use

* [Themes](https://docs.textpattern.com/administration/themes-panel) panel
* [Front-end themes](https://docs.textpattern.com/themes/front-end-themes) structure

## Notes

[This thread](https://forum.textpattern.com/viewtopic.php?id=49502) suggests theme creation is easier if you're running in flat-file mode with [etc_flat](https://forum.textpattern.io/viewtopic.php?pid=310108#p310108). But it can be done with just core functionality too? Does it warrant having two docs; one for core-approach and one with plugin approach? Would the differences be enough to warrant that?

### Cleanup

We should be able to eliminate most files in the /themes directory when this tutorial is written, excepting a similar doc for creating a back-end theme.

**Comment by wion** [comment]
I think the existing [Front-end themes](https://docs.textpattern.com/themes/front-end-themes) page can be used and refactored into a fully elaborated 'Creating a theme' tutorial.

**Proposal**: We don't refer to front-side themes as 'front-side' (i.e. we drop the adjective). We simply say 'themes', which means front-side by default. Back-end themes are distinguished by saying 'administration theme(s)' (i.e. using the adjective).

**Comment by wion** [comment]
#130

**Comment by wion** [comment]
Possible title change, or merge of page altogether. Well see how it plays out first...
https://forum.textpattern.com/viewtopic.php?pid=320264#p320264

**Comment by wion** [comment]
Title not yet actually changed, so correcting issue title to match status.

**Comment by wion** [comment]
@Bloke or @bloatware, in this current doc, it talks about the manifest.json file. It shows the file having these values if only the two required fields (name and title) are used when creating the theme:

```json
{
"title": "The Human-friendly Title",
  "version": "unknown",
  "description": "none",
  "author": "unknown",
  "author_uri": "none",
  "txp-type": "textpattern-theme"
  }
```
But when I run a theme creation test in the demo site and skip the optional fields, I notice Txp automatically adds `0.0.1` and the username as values for `version` and `author`. 

Wouldn't those values be what shows in a manifest.json file if the two fields were skipped? Or are the values still still just `none` as indicated?

**Comment by wion** [comment]
@Bloke or @bloatware, another question about the page (which I am currently adding to and revising)... 

It has a few sections that talk about default theme structure, what subdirectories and templates are essential or not in a theme package, and what Txp does if any essential part is missing when importing a theme. For example is says the pages directory and specifically the two files, _default.txp_ and _error_default.txp_, are essential and will be created if missing when importing a theme. So far so good understood. ;)

I'm a little confused about the nuances of this part, though:

> Theme updates thereafter will not overwrite the existing essential pages.

1. If these are essential elements, how is it possible that a theme would be distributed without them in the package? Wouldn't that mean the theme didn't work to begin with and shoudn't be shared?
2. What does 'theme update' mean? Does it mean to suggest like with plugins, where the original theme author creates a new version, and the current site owner imports the new version over the top of the older version?
3. If Txp initially auto-adds any missing essential elements when importing a theme, then those added elements must be used in the current site structure if the theme is used, right? Even if just the default template code?

I guess I'm confused why anyone would possibly share a theme that was missing essential elements, or that it's possible to even do so with Txp. That seems like something Txp should prevent from happening at time of trying to 'Export to disk', like throwing a error message saying: 

> Sorry. No can do. Your theme package is missing these essential elements: tack, tack, tack...

**Comment by bloatware** [comment]
Dunno, @Bloke and @NicolasGraph are your men for everything theme-related. But how can you upload a theme to @petecooper demo site, I thought it was write-protected?

**Comment by petecooper** [comment]
>But how can you upload a theme to @petecooper demo site, I thought it was write-protected?

It is. Well, it should be. I haven't found a way to bypass it from the front side yet. Obviously new pages / forms / styles can be added, but a theme cannot be added…now I'm worried.

**Comment by wion** [comment]
I'm not uploading there, just creating an initial them in the Themes panel using the New theme form.

**Comment by petecooper** [comment]
Ah, yeah - that's doable.

**Comment by wion** [comment]
My questions are with regard to what's written on the current page this issue is concerned with. Probably questions for @Bloke. Sorry @bloatware .

**Comment by Bloke** [comment]
> Theme updates thereafter will not overwrite the existing essential pages.

Outdated information. You can choose now via the checkbox when you import/export/delete to overwrite the files or the currently installed theme assets in the database. The first part of that section is fine: the stuff will be created if it doesn't exist (though it _should_).

The only time stuff might not exist is if the author has tinkered with the theme or built it on-disk from scratch (i.e. not used the admin interface to create the theme). The theme author may have renamed the 'default' Page as 'supergrid' for example and distributed the package like that. But on import, you will have the 'essential' Pages created (empty) just so things have less chance of breaking tags. 

We should probably review this in 4.8, since the new dev theme workflow means there's no need for the "essential" Pages to exist any more. Switching to a theme in dev will set the correct Page properly, whereas in 4.7 it is a "soft" switch that relied on the Page name being the same as in the live theme.

The concept of essential Forms, however, remains for now.

> [manifest.json] Wouldn't those values be what shows in a manifest.json file if these two fields were skipped? 

Yes, the version and author are auto-created if omitted.

**Comment by wion** [comment]
Great. Thanks, @Bloke. I can finish up this revision with that, commit, and then ping you for a technical review, if that's okay.

**Comment by wion** [comment]
@Bloke , I'm getting close to a finished first draft. I have another question, though, in the final stretch. Let's say you have the scenario where a theme change objective is to use the same site structure but a different theme on each section. As I understand it so far, one would then go to the Sections panel to individually assign themes to panels in the **Edit section** controls, respectively.

But, is that it in this scenario? I.e. if one does that:

1. Is it necessary to make additional new assignments in the **Uses page** and **Uses style** controls too? (Maybe only when the page and/or style name is different?)
2. Or, ideally, do they change automatically in relation to the theme assignment?†
3. Or does one have to make assignment changes in all three controls: **Uses theme**, **Uses page**, **Uses style**?

Any other relevant clarity appreciated. 

† Ideal, if true. And would suggest a possible UI change to better show the subordination of page and style controls to theme control by indenting them under the latter; making the power of influence more clear. Whatever.

**Comment by wion** [comment]
Now that I ask that second question and think about it, one probably has to make all three assignments, and the page and style assignment makes clear which of each in the theme package is the one to use?

**Comment by Bloke** [comment]
Yes, you *must* assign all three elements. When you alter the first drop-down (theme) the other two dropdowns alongside change to reflect the available Pages and Styles in that Theme. You can only pick from those assets available in the chosen theme you elect to apply to one or more selected Sections.

It changes very subtly in 4.8 - you can bulk apply a theme directly from the Themes panel and it will attempt to set the chosen theme (after confirmation if you want it to become Live) for all Sections that have matching assets.

For example, if your live 'About' and 'Portfolio' Sections used a Page template called 'one-pager' and your in-development theme didn't have that Page name (perhaps it only had 'default' and 'archive' defined) then clicking the 'Activate' link on the Themes panel (and confirming) would switch all sections that had either 'default' or 'archive' Page templates to the new theme. About and Portfolio would be left alone for you to manually (re)assign at your leisure using the multi-edit tool.

This workflow helps alleviate site breakage when switching themes, while also offering a handy shortcut for bulk-assigning Themes without having to do the multi-edit select-and-apply dance.

But, as I say, the latter is 4.8 functionality.

**Comment by wion** [comment]
Understood. Crystal. Thank you!

Erm, I've actually taken the approach on this, due to recent signs of a New Year rollout of 4.8, to center the doc around 4.8, so that Active feature is accounted for. :} But the doc is not so radically different that we can't fluff it hear and there with 4.7 notes (a la `{: footnotes .information} styling) to make the masses happy; the notes can then be excised at launch. I'll bank on your sound review comments for tweaking.

I will say here in advance that I'm really on the fence about this just becoming the new Themes panel because it's heavily centered on that panel. On the other hand, it covers related features and workflows from other panels (case in point) so it is outside the scope of just the Themes panel. And the delivery is a big departure from how panel docs have been written up to now; more explanatory and stepwise-progressive than just describing thing 1, thing 2, etc. And, it's the first test run of using images again in very minimal but strategic fashion. It is, quite literally, a tutorial, but entirely focused on core functionality; no hypothetical site structures or design examples. I like it. So, do let me know how to jump off this fence post! And if it does go to Themes panel, then it opens up the door for doing the same on other panel pages. But, there again is a fence-post quandary, because I think themes functionality is about the only panel page that warrants this kind of elaboration. I mean, I can't see doing this with languages, or preferences, lol, or hell, take your pick. But _Themes!_ After writing this, I see Textpattern back-end by a completely different model, with themes center stage. '_Just write!_' got the boot!

**Comment by wion** [comment]
I'm going fold in your insights on the last question, then commit. It won't be done, I'll have many passes to make undoubtedly, but it will be easier to see and tag-team on.

**Comment by wion** [comment]
@Bloke , edits so far are pushed to docs. Still a few sections to fill, copy to clean, images to change, etc. And I'm going to fold the import and default structure parts at the end into each other when I get to those. But there's enough there that you could look at the technical aspects of other sections, when you get a chance, and let me know if something is off or could use a bit more clarification. Add notes here, if any, by what section of doc they're concerned with, and I'll rope them in as I go. Thanks!

**Comment by wion** [comment]
@bloke, a near if not last question, hopefully.

I get the Update from disk scenario, which I presume means:

1. Move an new version of a community theme (which has already been imported to themes panel before) to the _themes_ directory
2. Use the **With selected** controls to update the database version.
3. Use the option for aligning/deleting files

But I don't understand how the community theme was added to the panel table to begin with. I.e. how do you first import a community theme? I don't see the obvious.

**Comment by Bloke** [comment]
Okay, there's a sort of symbiotic relationship between _disk_ and _database_ but live and in-dev themes can only be previewed on the site if they are installed to the database. The assets (Pages/Styles/Forms) won't run from disk... although of course you can reference anything in the folder so if you have additional js/css/whatever files in there, you can incorporate them directly.

To get a theme into play first:

* Download it from some repository.
* Unzip it to your '/themes' directory into the theme's own sub-directory.
* Visit the Themes panel.
* Above the table is a dropdown showing a list of themes that are _on disk but not installed in the database_. Pick the one you just unzipped.
* Click Import.

Then it's in play and you can make changes in the DB (live/dev), make it live, whatever. You can 'backup' your theme assets to disk and/or make changes on disk and import those to your database so they can be used on your site.

If you subsequently delete the theme from your database _and elect to leave the disk files alone_ then the disk-based theme once again appears in the dropdown above the table and can be reimported. Only when all themes are in the database will the dropdown not be present.

Does that make sense?

**Comment by wion** [comment]
Yep. I think so. This is the part I was missing...

> Above the table is a dropdown showing a list of themes that are on disk but not installed in the database.

This 'disk menu' doesn't appear in the demo site, of course, so it must be a contextual menu that only shows up when there are themes on disk?

**Comment by wion** [comment]
> Only when all themes represented in the filesystem are in the database will the dropdown not be present.

Heh. We're good!

**Comment by wion** [comment]
@bloke, or anyone, I'm not in a position to 'import' themes, so I have no idea what the theme import menu looks like that appears at the top of the Themes panel table when an external theme has been initially added to the _themes_ directory. Can anyone please provide me with an image of the menu itself, open, and showing one theme named and titled as follows, whichever is the case:

* xyz-some-other-theme
* An External Theme

This sample naming fits the opposite of the default new theme creation examples being use, which are:

* abc-my-new-theme
* My New Theme

Ideally, you'd provide two images: One of the menu closed, as it would appear normally, and one of the menu open showing the 'xyz' theme as  named/titled above, whichever accurately reflects. I'm guessing it's the Title (ideally).

If you look at the doc, you'll see that the convention I'm using is to crop images directly to the borders of the actual interface controls, so that is shows the borders and control label, but not a pixel more.

And, if you want, push them to _/img_ directory using the following names:

* import-theme-menu.png
* import-theme-menu-select.png

🙏

**Comment by wion** [comment]
@wion,

Finish the following sections and first draft is done:

* Sections preview
* Updating imported themes
* Removing themes

Then just some editing passes and images fine-tuning.

**Comment by wion** [comment]
Title and file name has been changed to something more reflective _and_ attractive.

**Comment by wion** [comment]
@Bloke, one last question, I think. I'm trying tie the sections about default package contents and Txp's handling of external themes when they lack essential default elements, etc. to the importing or updating sections, whichever is relevant, or if both are. My confusion at this point is which is relevant. I think I made it look/sound like the package augmenting is tied to the **importing** of external themes (that's what made sense to me since it was the 'first contact' with database, as it were), but now I'm thinking it's the **updating**  process? That was a term I remember seeing a lot, in fact. Can you clarify? I can then readjust those sections and context, revise the copy, etc, and be able to finish this p for now.

**Comment by Bloke** [comment]
This is how it works if the on-disk Theme contains 'essential' assets that are missing:

* If you Install an on-disk theme from the dropdown above the table, all Theme assets (Pages/Forms/Styles) are imported and any missing 'essential' ones are created, and populated with a placeholder HTML comment.
* If you subsequently delete such assets from the database, that's your prerogative.
* If you then export to disk, your theme on disk will *not* have the essential items you've deleted in it. If you zip and distribute it like that, fine.
* Anyone - including yourself - who updates a theme from disk that contains missing elements will have those assets created automatically in the database. Every time. Placeholder content will be injected in those assets if they were missing on Update.
* If you delete them and Update from disk again, they will be recreated in the database.

Now, that's as it works _today_. I'm seriously considering - and I'd like your thoughts on this @bloatware - whether to relax this situation in 4.8 now we have a proper live/dev split. I don't think the `default` Page and Style is necessary any more. When anyone assigns a Theme to a Section, they MUST choose a Theme and MUST choose (from it) a Page and MUST choose (from it) a Style.

We're now protecting content when switching via the Active/Preview links on the Themes panel. For example, if a switch to any given Section would result in that Section not being assigned a Page/Style due to a missing asset compared to how it was before, the Section is skipped for you to do manually later.

I don't think there's any reason to keep this notion of enforced 'default' for Pages and Styles.

However, I think an enforced `error_default` might still be valid. What's the impact if that is missing? What does Txp do if it finds no error path? If it relies on this, could we design it such that it doesn't?

`error_default` acts as a backstop to catch all error situations and displays a page to that effect with the code/status message. If you have a dedicated `error_404` or `403` or `500` or `451` or whatever, then those are checked first and the more specific one rendered instead. If after checking all that it came up empty, is there something we could do? Perhaps render a bare-bones HTML page like we do if the database is missing? Just status code and message, nothing more.

If we did that, we then only have Forms that require this 'essential' treatment due to our ingrained behaviour inside some tags.

Is that workable or have I missed something? I know we hard-code 'default' in core as a fallback Page/Style, but is there any need? Can a Page ever be assigned "nothing" even by mistake? What would be the effect of not having any hard-coded 'default' mentioned? White screen of death perhaps (in the case of Page)? Unstyled content (in the case of Style)? In which case, we probably need to leave this convention intact or find a way to handle it more gracefully.

I'm all for convention, but the above workflow does seem a little OTT. Thoughts?

**Comment by bloatware** [comment]
> I don't think the default Page and Style is necessary any more.

I was already for relaxing it in 4.7, but which Page/Style would we use on sectionless "home" page?

**Comment by Bloke** [comment]
Oooh good point. Ummm. Yeah. Maybe we need it.

**Comment by bloatware** [comment]
Though `default` section has its Page/Style assigned too, my bad. You are right, `default` Page/Style are only default, not mandatory.

**Comment by Bloke** [comment]
Ah yeah, true. Just wondering if we can remove the hard-coded 'default' since I can't see any situation nowadays where a Section will ever be "unset". Maybe I'm not thinking hard enough.

**Comment by Bloke** [comment]
Great stuff, Destry. Really lovely tone to it.

Notes/edits (suggested changes in bold):

* The single directory can be thought of as a ‘package’ and its contents can be thought of **as** ‘assets’ organized in subdirectories by type.
* You can switch theme context again at any time from any assets panel by using the **Theme** menu again
* assets between the different themes are paired **well**
* Future updates of the imported theme (i.e. after the theme author releases a new version) **will** overwrite the elements Textpattern added. [updating a theme from disk creates an exact copy in the database from what is on disk, merging with what you have already and overwriting if any assets are named the same. Optionally, if you check the 'Delete unused templates' box, any assets in the theme in your database that are not on disk will be deleted.]

The "Assigning multiple themes" section needs some light editing, methinx. The workflow here hasn't changed much since 4.7. It's always been integrated such that when you pick a theme, the Pages and Styles list is populated with the assets in that theme only.

Also need to mention that instead of assigning them one by one, you can use the multi-edit tool. That invalidates: "For this situation, you must manually assign themes, and the relevant page and style in the theme package, to each section, one at a time..." because you can apply a theme's Page/Style to multiple Sections at once via multi-edit.

The main difference in 4.8 is that the dev/live switch in the multi-edit tool works differently. Instead of being a 'soft switch' that tries to assign the section as a preview - and only if it can - it now makes a dedicated switch to the chosen assets. The dev and live environments are distinct and well-defined (there's a column for the dev theme assets in the txp_section table, instead of just relying on a pref like it did in 4.7).

When exporting, we'll probably insist that a theme is zipped up for easy distribution and we'll only host a single downloadable bundle file per theme on the Themes repo, so you might wanna also mention that once you've exported a theme, to zip up the abc-my-theme directory and everything below it for distribution.

Finally, for now, are we standardizing on using 'form' as a name for an 'edit panel'? Seems a mite confusing given we're talking about Textpattern _forms_ too. We standardized on using 'panel' (e.g. _The Preferences panel_, _The Write panel_, etc) to disambiguate from being termed a Page and thus clashing with our template Page concept. Should we come up with a disambiguated name here too for the 'edit' step?

Thanks for all your hard work on this. Awesome stuff.

**Comment by wion** [comment]
> - [x] The single directory can be thought of as a ‘package’ and its contents can be thought of as ‘assets’ organized in subdirectories by type.
> - [x] You can switch theme context again at any time from any assets panel by using the Theme menu again
> - [x] assets between the different themes are paired well

This one I couldn't find based on your text, which means I had probably edited out the former copy since you first looked at it. But I could tell you were talking about the 'Updating themes' section, so I added a version of that as the second para in that section. It's a nice addition. Thanks!
> - [x] . . . [updating a theme from disk creates an exact copy in the database from what is on disk, merging with what you have already and overwriting if any assets are named the same.

Btw, in relation to the above item, the checkbox option on 'Update from disk'; the string reads: 'Delete theme templates from disk too'. It's not phrased well, and missing punctuation. Better like this: ‘**And delete theme assets from disk?**’ It's a little different from the others because the _total_ delete action is already a given on database, now we're asking '_And_ from disk too?'

This next one, I never knew that about the contextual behaviour of pages/styles menus to the themes menu. But then I've never created a theme before either. ;) I'm sure this is all foreign to most people, though. Anyway, I suspect, again, I had revised this section to account for this part, but please verify let me know if not yet right.
> - [x] The "Assigning multiple themes" section needs some light editing, methinx. The workflow here hasn't changed much since 4.7. It's always been integrated such that when you pick a theme, the Pages and Styles list is populated with the assets in that theme only.

This part confused me with the 'multi-edit' tool. But I see now, you mean the **With selected...** controls, then select 'Change theme/page/style'? Good catch. I wasn't aware of those. Yes, that's easier than one-by-one as a first go-to. I'll work that in. Thanks!
> - [x] Also need to mention that instead of assigning them one by one, you can use the multi-edit tool. That invalidates: "For this situation, you must manually assign themes, and the relevant page and style in the theme package, to each section, one at a time..." because you can apply a theme's Page/Style to multiple Sections at once via multi-edit.

Not sure I want to get too unnecessarily and specifically contrasting between hard and soft switching on this; goes over the new user heads (and mine, a wee bit still)...
> - [x] The main difference in 4.8 is that the dev/live switch in the multi-edit tool works differently. Instead of being a 'soft switch' that tries to assign the section as a preview - and only if it can - it now makes a dedicated switch to the chosen assets.

Same here on the checkbox options... You can set both Dev and Live and the same time? What's the scenario for wanting to do that?
> - [x] The dev and live environments are distinct and well-defined (there's a column for the dev theme assets in the txp_section table, instead of just relying on a pref like it did in 4.7).

In relation to that section assignment option, though; the whole line of boxes and menus; it might be easier to take in lists form, no?. See my image for that one, which was kind of necessary anyway to make work as a sane doc image width.

Have done. But is there a particular zip format wanted? Or will it matter?
> - [x] When exporting, we'll probably insist that a theme is zipped up for easy distribution and we'll only host a single downloadable bundle file per theme on the Themes repo, so you might wanna also mention that once you've exported a theme, to zip up the abc-my-theme directory and everything below it for distribution.

> are we standardizing on using 'form' as a name for an 'edit panel'? Seems a mite confusing given we're talking about Textpattern forms too . . . Should we come up with a disambiguated name here too for the 'edit' step?

Ah, that old chestnut. I was just using 'form' in a normal 'web form' sense, not really thinking about it. I could modify such references to things like '**Edit theme** form**', where the label, normally a proper noun, becomes a proper adjective. Perfectly fine. I don't think standardizing any specific term will work though. And who will pay attention anyway? ;) 

A good rule of thumb is (worth adding to the guide, in fact), if the grammar/spelling is correct, strings are formatted correctly (within best reason), _and_ it just makes sense at the sentence level. It's probably fine. Each para might need its own jiggling, but that option to just find the right syntax/diction is often good enough and leeway enough.

### Still to do

- [ ] Add [Sections preview](https://docs.textpattern.com/themes/themes-a-complete-walk-through#sections-preview) section.
- [ ] Add missing image for [Switching themes](https://docs.textpattern.com/themes/themes-a-complete-walk-through#switching-themes) section, which I'm holding off on until I see if the blue 'Active' links will actually become 'Activate', as discussed in the cms repo on that issue.
- [ ] Add missing images in the [Importing themes](https://docs.textpattern.com/themes/themes-a-complete-walk-through#importing-themes) section. But I can't do those because I don't currently have a 4.8-dev install. I need someone to give me some screenshots of what the Import menu controls look like when closed and open (selecting a theme).

**Comment by wion** [comment]
It was late last  night when working on the 'Assigning multiple themes' section. Is the Irish thing too  much. I kind of like it.

**Comment by wion** [comment]
@Bloke , is what is said here about the theme 'shell' remaining, along with the non-standard elements? Or does Textpattern actually know how to remove the package directory too and leave only the non-standard files sitting there like abandoned animals, but not showing up in the Import menu?

**Comment by Bloke** [comment]
The 'shell' thing is perfect.

**Comment by wion** [comment]
I can't find the Issue, but I thought @bloatware had mentioned somewhere that blue 'Active' links would be changed to 'Activate' links, which is why I'm leaving that for now. Is that actually the plan, though?

**Comment by wion** [comment]
I thought I might add one more section, very end; something like 'What now?' and briefly give a vew pointers toward the actual design part of it. Maybe flat-file plugins, an eye on Textpattern Tips, whatever. Any ideas to throw in there? The point being: 'now you have mechanics; here are some creative approaches...' kind of thing. I don't know.

**Comment by bloatware** [comment]
@wion I've just pushed `activate` to dev, but would strongly suggest that we switch to pictograms whenever possible. Multiplying text strings is bad for performance, and they are not easily reusable in all languages. For example, the current `active` string does not agree well with 'theme' in Russian.

**Comment by wion** [comment]
> . . . strongly suggest that we switch to pictograms whenever possible.

No problem for me. I guess that's a @philwareham thing. Can you provide that issue link here, where that discussion was happening, please? I'll try to follow along, but someone should remind me if/when the presentation of those links changes again.

By the way, is it a problem to make the green 'active' links bold, too. I think that really helps with the distinguishing aspect under the current conditions. If not, no big deal.

**Comment by Bloke** [comment]
Good stuff, thanks. I'll read it more thoroughly later. A few things:

* _Importing themes_ section: "presumably downloaded from a **shared** location".
* The second para in _Updating themes_ should probably end with: "overwriting existing ones having the same asset name." Since using 'file name' implies a file extension.
* The 'described earlier' link destination in the third para in the same section needs a trailing 's': `#duplicate-via-the-with-selected-controls`.
* Not sure about the Irish thing! Doesn't do anything for me either way. And, btw, is green/blue color blindness a thing? Should we be solely relying on shade to differentiate something? I suppose once it changes from 'Active' to 'Activate' then that's more visually different than colour alone.

Might be worth an aside to mention you don't _have_ to use disk at all for themes. It's only if you want to share them, or back them up, or put your theme assets under version control that people will likely do it.

A further note on that point: if you create a new theme in the UI it will *not* appear on disk. It's solely resident in the database until such time as you Export to disk. Now, if you choose not to export it at all, the checkboxes in the _With selected_ tool regarding 'disk' operations don't make sense. What disk copy?

Unfortunately we can't omit the checkboxes, as you may choose one or more themes from the list and some might be on disk, while others might not. JavaScript does not know and can't find out directly.

We could, perhaps, 'tag' each row in the UI with a `data-` attribute indicating its on-disk status. Then perhaps if you select all entries that are not represented on disk, the checkbox could be omitted? Not sure if that is worth investigating or if it makes things too confusing from a UX/documentation viewpoint. (@bloatware what do you think?)

Either way, if you try to perform a disk-based operation (e.g. _Update from disk_ when the theme is not resident on disk) you'll get a warning after the action runs telling you which parts it could or couldn't do. If the entire disk action fails you'll get a red failure box. If part of the action fails (e.g. it could update one of your choices but not the other) you'll get an amber warning box. If all actions succeed you'll get a green success box.

> ‘**And delete theme assets from disk?**’

Fine by me. Or fix the punctuation. It's just a language string.

> I don't really talk about 'soft switch'.

Don't need to mention the 'soft' switch nature of 4.7. That was just for you, for context.

> Maybe needs expansion... You can set both Dev and Live and the same time?

Making your live and dev environments match avoids surprises, since you'll see the same theme when you browse the front end, regardless if you're logged in or not.

> In relation to that section assignment option, though; the whole line of boxes and menus; it might be easier to take in lists form, no?

Yes, I did wonder that. If we put line breaks in or divs or something around the two blocks, I presume JS will render them that way when it draws that fragment, but I don't know about lining things up. @philwareham what do you think to something like this?

[mockup change theme/page/style](https://docs.textpattern.com/img/select-change-theme-option.png)

> is there a particular zip format wanted? Or will it matter?

Nope. Out of our control at the moment. Txp doesn't handle zipping/unzipping yet. Perhaps if we allow you to "upload" a theme bundle when the Themes site comes online and unpack it on disk, we might care. But that's only a possibility.

**Comment by bloatware** [comment]
>  is it a problem to make the green 'active' links bold, too.

No problem for me. I guess that's a @philwareham thing. :-)

**Comment by Bloke** [comment]
Can we tweak the Importing themes number list for clarity please. Something like this (tweak to taste):

------

If you didn’t see the menu before, it is because it works in context with the themes directory, as follows:

1. **One or more disk themes with no database equivalents**. When one or more themes reside in the *themes* directory that are not in the database, the selection menu appears above the themes table, permitting the unrepresented themes to be imported to the database. The menu is not visible if this condition is false.
2. **Disk themes match database themes**. When the same themes are equally represented in the database and on disk, or only in the database and none on disk, the contextual menu is hidden, since there is nothing to potentially import.
3. **Database theme deleted**. If/when you delete a theme from the database but not the disk-based version (assuming there is one), the theme import menu will appear again, as described in condition 1 above, because the first condition becomes true.

**Comment by wion** [comment]
FYI, just focus on the developmental editing for now. Typos and other copy edits can wait. I might even find those along the way anyway (such as that link). You can even do those when I'm done with the final first draft. ;)

**Comment by wion** [comment]
> A few things:

Those are all done. I removed the shamrock nonsense. Like I said, working late. ;)

> Might be worth an aside to mention you don't have to use disk at all for themes.

Good one. I've added a parenthetical. And see the endnote that links back to it. Easier. ;)

> A further note on that point: if you create a new theme in the UI it will not appear on disk. It's solely resident in the database until such time as you Export to disk. 

Yeah, I think this is intuitive from the rest. At least that's how I understood it at this point, and if I understand, it's dummy-proof.

Cont...

**Comment by wion** [comment]
> [mockup change theme/page/style](https://docs.textpattern.com/img/select-delete-option.png)

That's not to suggest the design, exactly; just to show an alternative idea to linear. To me it's a lot easier to make sense of.

**Comment by wion** [comment]
> Now, if you choose not to export it at all, the checkboxes in the With selected tool regarding 'disk' operations don't make sense. What disk copy? . . . Unfortunately we can't omit the checkboxes, as you may choose one or more themes from the list and some might be on disk, while others might not. JavaScript does not know and can't find out directly.

Hmm... I know you're just talking aloud to me, but that might actually deserve being an endnote to provide some explanation for those who might come across it. I don't want to weigh main text down with such side explanations, though.

> Either way, if you try to perform a disk-based operation (e.g. Update from disk when the theme is not resident on disk) you'll get a warning after the action runs telling you which parts it could or couldn't do. If the entire disk action fails you'll get a red failure box. If part of the action fails (e.g. it could update one of your choices but not the other) you'll get an amber warning box. If all actions succeed you'll get a green success box.

That, OTOH, is probably worthy to add in text somewhere. I'll get back to that later.

**Comment by wion** [comment]
> Can we adjust the Importing themes number list for clarity please.

Yep. Switched out. Thanks. That was just figuring aloud on my part.

**Comment by wion** [comment]
> Making your live and dev environments match avoids surprises, since you'll see the same theme when you browse the front end, regardless if you're logged in or not.

This one I might need more clarification... Let's look at this from the context of someone who just installed Txp, is not a developer. Might only want to swap a 'skin' on the default theme. What does 'live' and 'dev' environments mean in a default install of Txp?

**Comment by Bloke** [comment]
> 'Development' is that new location stored in the txp-tables, or something? And 'Live' just refers to the websites live front-end?

'Development' is a theme that is active on one or more Sections that are _only for logged-in users to see_. That is, a preview of what the changes you've made will look like. This allows you to see the effect of your meddling without affecting public visitors (as long as you're careful and don't start deleting plugins or changing prefs or whatever).

'Live' is what everyone else - the public, not logged in - sees.

I admit that setting them both the same is not a usual occurrence but if you're done with your edits and want to synchronise everything for now until such time as you start diverging with a new theme in future, it's handy to be able to do it in one step instead of having to set the current theme live and then go back and do it again to change your dev environment.

**Comment by wion** [comment]
That last share was helpful in two respects: in addressing that relevant section better about the two options (Dev/Live), etc. and segueing into starting the Previewing section. 

I realize now what was nagging me about the current document structure, too, so I'm going to slightly change the headings/structure in that respect too.

**Comment by wion** [comment]
@Bloke, in the re-headed [Previewing (development) themes](https://docs.textpattern.com/themes/themes-a-complete-walk-through#previewing-development-themes) section, I think I should add an item (to the list of things to especially pay attention) about the Themes column. But what should I say? I see the grayed out default theme name and a new name next to it. What is that telling me?

**Comment by Bloke** [comment]
The greyed-out one is the one that's publicly live. The other one is the dev one that overrides it. Thus it shows visually which is 'active' for you and other logged-in users, while also showing you what the rest of the world see.

You may also see some assets in red. Those are where you've switched to a theme (perhaps using _Developer preview_ from the With selected... tool, or when using Active/Preview from the Themes panel) but the page/style that was in place has been deleted or is not present in the theme. That's a warning that if you make this theme live as-is, things will break as you will not have the red assets assigned to it.

**Comment by wion** [comment]
@Bloke, on that mention of 'Developer preview'... what is the difference between the two options?

* Deploy to live
* Reset to live

Also, here's a test I just tried on the demo site...

1. I duplicated a theme, then went to the Pages panel and deleted the 'archive' page for it (the red "x" delte icon was sho. Then went back to Themes panel to preview the duplicated theme. In the preview context, it showed the red text as you described for the missing archive page that is/was assigned to the 'articles' sections. I assume the greyed-out name was the live template functioning under the overriding context.
2. I then checked the box for that section and used the 'Developer preview' function mentioned above, and ran the 'Reset to live'. It seemed to bring the greyed-out (live) page template assignment to the surface, in context of the overriding preview view. (I kind of get that, it's applying the live/working page where you've said you don't want an alternate from the development theme's assets. Okay for your own site. Probably not a good idea for a theme to be shared, right?)
3. But then I tried to get back to original conditions, before deleting the theme page template and seeing the red text. I can't do it. That's not to say it's not possible, but I can't figure it out.

So, I guess my questions are:

* What is the 'Reset to live' option really for? (And did I not make a wise choice selecting it in scenario 2?)
* In scenario 3, how would I get back to original conditions, before seeing the red text? Use both options at once (Deploy to live / Reset to live)? Delete the duplicated theme and start over? 

There are so many ways and places to change things now, and contexts, that I get overwhelmed and lost in the woods if I go too deep into them. Can't find my way out again.

**Comment by wion** [comment]
The point of asking is that I'm explaining the red text scenario as you mentioned before, and trying to offer options what to do in such a case? That last scenario is confusing me. See the other two mentions in that section...
https://docs.textpattern.com/themes/themes-a-complete-walk-through#preview-link

**Comment by Bloke** [comment]
Developer preview allows yo to decide what you want to do with the current 'preview' or 'staged' (in-development) theme assets that are noted in the section(s) you selected. So:

* _Deploy to live_: will push the current _development_ assets to the live site for each selected Section, replacing any previous live assets that were employed.
* _Reset to live_: will retain the current _live_ assets that are in force for each Section, and remove any previous development assets that you (or other logged-in users) may have been using to preview the select Sections. In other words, remove any development overrides, so logged-in users see the same as live visitors for the chosen Sections.

Yes there are many ways to achieve things here, but if you think about it from a workflow perspective, you can take action is a variety of ways. For example:

1. **I want to assign _the same assets_ (live and/or for logged-in-users only) to a bunch of Sections** Use the With selected... tool and choose Change theme/page/style.
2. **I want to assign _the same assets_ (for logged-in-users only) to a bunch of Sections so I can take a peek at my edits and tweak the site design before deploying** Use the With selected... tool and choose Change theme/page/style, along with the Development checkbox.
3. **Okay, I've got a bunch of development templates set up that I like - different ones from different themes assigned to a bunch of Sections - and want to make them live in one swoop** Select the sections and use _Developer preview_ with the _Deploy to live_ option. Without this feature, you would have to use the _Change theme/page/style_ multiple times; once for each group of sections that shared the same assets. So _Deploy to live_ acts as a shortcut that allows you to 'prep' the sections to how you want them (in dev) and then make them live in one shot, regardless of which themes and assets are in use on each Section.
4. **I've installed this kick-ass theme, and want top preview it on my entire site** From the Themes panel, click 'Preview' against the theme you want.
5. **I've installed this kick-ass theme, had a look at it, love it and want to push it to my entire site** From the Themes panel, click 'Activate' against the theme you want, and confirm. Or, if you've already staged it on some or all sections, you can select some/all sections and use _Developer Preview->Deploy to live_ instead.

So as you can see, depending on what you want to achieve there are a few routes. All of them lead to the same place - assets assigned to sections for you and/or logged-in users - it's just that some are more efficient than others, involving fewer clicks and hoops, depending on what you want to do.

The bottom line is that we're allowing your live server to act as a staging server as well. So you can develop and preview and have clients sign-off a design in (relative) safety _using live article content_. No other CMS (as far as I'm aware) does this. You usually end up:

a) exporting the database and files, setting up a new (sub)domain or server with a Txp install containing the cloned content, make your changes, then have to worry about porting the changes back and how to resync the databases after your live content has moved on since you took the DB snapshot.

b) setting up fake Sections and assigning your dev templates/styles there, to allow clients to see what you're developing... but then you have to create a bunch of fake content too, or muck about with <txp:article_custom>, since articles can only be assigned to one Section at a time.

Our dev/lve workflow tools sidestep all the hacks and all the drudgery of the staging/live dance.

It's pretty fuckin' sweet :)

**Comment by wion** [comment]
Excellent. Yep. That's how I like to approach it, by task objective and the most efficient way to complete it. It's for that reason I don't think every path to the doghouse needs documented. Just the most obvious/quickest/intended paths, if that makes sense. But my paths may not always be the paths others like. And in this case I'm still discovering this functionality. So if I seem like I"m going astray in the doc, let me know. Like you've done here. Thank you!

**Comment by wion** [comment]
@philwareham, @bloke, @bloatware, 

Are we pretty stable with the Themes UI design now? I'd like to get back to this doc and revise/finish it in relation.

**Comment by philwareham** [comment]
I think so, I’m happy with the latest UI. Definitely finished for 4.8.0 release anyway. Cheers Destry - appreciate your patience on this documentation with all changes we did!

**Comment by wion** [comment]
Okay. Revising against the latest 4.80-dev demo.

**Comment by bloatware** [comment]
What should we do with the former 'dev preview' link/functionality? Dump it entirely (and remove its code), put the link on individual theme page or something else?

**Comment by wion** [comment]
I'm just trying to keep up, but I thought the 'dev preview' or 'preview' functionality _was_ dropped in favour of the Select with... functionality on the Section panel side. But let me see if I understand as it is now with a clean installation:

1. Go to Themes panel and see one (default) theme. No 'In use' indicator showing.
2. Click **Assign sections** link. (Seems to me sections are already assigned, so I go back to Themes panel.)
3. Click theme name. (I want to duplicate the theme to have a new development theme.)
4. Add new metadata for an 'abc-new-theme' and click Duplicate link. (Now I'm back in Themes panel and see two themes. The default theme finally appears with the 'In use' pill.)

To me at this point I have now created a theme in development. If it can be done before duplicating the default theme, it's not very obvious.

Now let's say I've made a few changes to the new development themes CSS file. How do I preview that? Do I click the **Assign sections** link for that theme? Is assigning sections what must be done before I can preview it? I don't know, but the command 'assign sections' may not make sense at this point if you came by my path.

What should I do to preview my dev changes at this point?

**Comment by philwareham** [comment]
I think we can drop the dev preview link idea, yes.

**Comment by bloatware** [comment]
The idea was to provide a one-click way to 'preview' in-dev themes, and @wion seems to agree it makes sense. It's a relatively safe operation, unlike assigning 'live' themes (replaced by 'Assign sections' link now). The question is where to put 'preview' link: individual or list themes page? Checkbox would not do because you can not preview multiple themes at once.

**Comment by wion** [comment]
> Checkbox would not do because you can not preview multiple themes at once.

Arg! I overlooked that.

What is an ‘individual’ themes page? You mean in sections panel context?

If so, then, as mentioned, I think the proposition of ‘assign sections’ comes too early. What if I want to create new sections first? In which case ‘assign’ doesn’t even make sense yet.

This makes sense to me:

1. Create/duplicate theme. (A new dev theme.)
2. Preview dev theme.
3. Or edit one or more of its assets.
4. If 3, then preview dev theme. (Decide if ready to use existing or create new sections.)
5. Use selection controls in dev preview mode to assign assets to sections. (Or create new sections first.)

In other words, I think it makes more sense to change the ‘Assign sections’ link on the Themes panel records to ‘Preview’, then leave the _assigning_ steps to the sections preview view, where the action is more in context with the appropriate panel (and decision flow).

If none of that makes sense, then I give up, because I still don’t understand it either, apparently.

**Comment by wion** [comment]
> Checkbox would not do because you can not preview multiple themes at once.

What if you walked them through it with an error message to make it clear they can only select one box in this case?

1. User ticks three boxes.
2. User selects 'Preview' from selection menu. 
3. Gray pop-up error is thrown saying 'You can only select one record for this option.'
4. User clicks 'Okay' to confirm.
5. User checks one box only and successfully previews dev theme.

**Comment by bloatware** [comment]
Individual theme page is where you land when clicking on 'New theme' or an existing theme link. But it's not ideal workflow-wise since before previewing a new theme one must save it, which takes him back to the themes list.

Restricting the multi-edit widget to a single item does not look very natural. And it takes two clicks anyway...

**Comment by wion** [comment]
> Individual theme page is where you land when clicking on 'New theme' or an existing theme link.

Ah, the metadata form? That's what it seems like to me since there's nothing there to do with what's actually in a theme (i.e. assets), just most of what goes in the manifest file.

Well, if that's the case, then I hardly see anything wrong with putting a Preview link inside the form view. Like you said, it's harmless; certainly won't make things worse.

**Comment by wion** [comment]
Let me know what y'all decide about the [dev theme Preview functionality](https://docs.textpattern.com/build/themes-creating-using-and-sharing#previewing-themes-in-development). I had already spent some time writing documentation about it, and when read that again, it seems like useful functionality indeed. I personally don't understand why you would not want development preview functionality. It seems essential in themes development.

**Comment by Bloke** [comment]
I still think it has value to be able to quick-switch all dev -> to live or to revert back to live.

**Comment by bloatware** [comment]
I'm also missing some kind of 'assigned sections' indicator for theme select box on Form/Page/Style tabs. You do some changes and then realize it's a wrong theme...

**Comment by wion** [comment]
Question...

Let's say I do this workflow:

1. Duplicate the default theme.
2. Click the 'Assign sections' link on the duplicated theme.
3. Assign the duplicated theme's default page to the default section, but for 'Development' only. (Then I see the orange dev pill on that section record.)

What am I looking at when I click the 'View' link under the Name column for that section now? Is it the live section or the new dev section?

If it's the new dev section I just created in step 3, then that is effectively the 'preview' feature I would be hunting for as a user, and we can probably not quibble further about it. Though I think that connection needs to be made more clear, case in point.

**Comment by wion** [comment]
Maybe those 'View' links should be in the Theme column instead, and immediately next to the dev or live theme, as the case is. Then the association is perfectly clear.

**Comment by bloatware** [comment]
> What am I looking at when I click the 'View' link under the Name column for that section now? Is it the live section or the new dev section?

Devs who have enabled 'dev preview' option are always seeing the 'dev' theme when one is set.

**Comment by wion** [comment]
'Dev preview' option?  What else am I missing? Is there a spec everyone except me is working off of?

**Comment by wion** [comment]
Do you just mean to say that when a 'Development' pill shows up on a section, then that's the trump view for that section?

**Comment by bloatware** [comment]
It depends on `Admin/Enable development theme preview?` preference state. You can enable or disable dev preview there.

**Comment by wion** [comment]
How did I keep overlooking that? No wonder I was clawing in the dark.

Okay, that changes things. So as it seems now, I can _preview_ a development theme (meaning I can view it on the front end as I edit assets). As far as it seems to me, that's all the preview functionality one really needs.

I'll leave it to you devs to decide if a quick theme switcher thing is needed, but as I see it, this functionality is already provided for by the selection controls in the Section panel after clicking the 'Assign sections' link. Maybe we should leave it at that and move on.

Unless I'm missing something, there is no further issue here, right? I think I grasp it now to finish docs as the current 4.80-dev demo shows.

**Comment by wion** [comment]
[Bump](https://github.com/textpattern/textpattern.github.io/issues/128#issuecomment-560007917)

I really need these images, but I can't produce them from the demo site. I assume you devs have tested the import features (thus see the menu) somewhere... Can anyone provide the two images as requested, _s'il vous plaît_.

@Bloke , @bloatware , @philwareham

**Comment by wion** [comment]
Maybe I'll see if I can find a theme to import in my own site. Any suggestions of one to try. I won't actually use it, per se, but at least trigger the menu.

**Comment by Bloke** [comment]
Sorry, @wion. I can take this if you like, but you can do it yourself on a site:

1. Visit the Themes panel.
2. Clone an existing theme.
3. Rename it to xyz-theme.
4. Use multi-edit to Export to disk.
5. Delete it from the Themes panel but ensure the 'delete from disk' checkbox is unchecked.
6. The import menu appears with your disk-theme in the list.

**Comment by wion** [comment]
Heh. I didn't even think about exporting. I'll do that, thanks, @Bloke. It would probably be good to walk through both process anyway.

**Comment by wion** [comment]
Easy-peasy. Thanks.

I'm not sure how many theme's I'll ever actually make, but I've certainly come to understand the themes functionality better after all this.

**Comment by wion** [comment]
@Bloke, I think I've updated this doc now so it's correct against current functionality, including imagery. When you get a chance, please give it a technical review.

Don't trouble yourself to keep a list of problems found, to relay to me, yadda yadda... Feel free to tweak and fix whatever issues you find, technical or orthographical.

When your done, I'll give it one last proof read, then seek debutante themers from the forum for their feedback on understanding the concepts explained. The real test.

**Comment by Bloke** [comment]
I'm about a third of the way through (slow on this phone usung H+ by the seaside!) and it's great so far.

One thing: there are a few internal links to 'assigning multiple themes' and that segment appears to have been axed. Should the links be pointing to another section I haven't found yet, or is this section meant to be in the doc somewhere?

Also, in the 'asset association' section, there's a broken image link. Can you fix that please? Ta!

**Comment by Bloke** [comment]
Gone through it all. Nice work, Destry-san.

**Comment by wion** [comment]
I'll fix the items you found. Thanks a lot.

**Comment by wion** [comment]
@Bloke, I've made a couple structural changes where the flow and bloat was bothering me.

For example, there's a new [Multi-selection controls](https://docs.textpattern.com/build/themes-creating-using-and-sharing#multi-selection-controls) section now that brings better attention to those important controls, specifically the 'Change themes/pages/styles' option. This also makes it possible to refer back to that section later in the doc to reduce repetitive descriptions where feasible.

I've also created the [Assigned sections indication](https://docs.textpattern.com/build/themes-creating-using-and-sharing#assigned-sections-indication) section to bring better focus to those new pill features too. In this case a couple of things occurred to me:

1. Sections panel: If it's now possible to have more than one theme name in a section record under the Theme column, shouldn't that column label now be plural, 'Themes'?
2. Since it's possible to assign a theme as both 'Development' and 'Live' Does that theme then appear in the Themes panel table with both the orange and green 'In use' pills on it?

I didn't actually try assigning both dev and live to a theme, so I'm wondering if that question is true, then I should probably add that graphic to this section too.

**Comment by wion** [comment]
Scratch that second question. I just tested it. ;)
It seems not to be the case; rather, if a theme has both a dev and live assignment, it simply appears with a green 'In use' pill and no orange indication at all, either in the Themes panel or Sections panel. 

I guess that makes sense, but certainly a behaviour to make clear in the doc, so I'll do that.

**Comment by Bloke** [comment]
Nice idea regarding the multi-edit section. That does consolidate things nicely.

Regarding singular vs plural column names, well, we have singular 'Page' and 'Style' too, yet there could be a dev and a live Page assigned separately (if, for example, you named a Page differently in your Dev theme compared to the Live theme, you see the horizontal line and the two Pages listed).

There's only really ever one Theme/Page/Style in use - the live one. The dev state is just a convenience for iterating a design, so it's probably not worth the upheaval to change the labels at the moment.

Oh, one more thing, where did mention of the Deploy to live / Reset to live multi-edit option go? It only appears if you have set a dev theme.

**Comment by wion** [comment]
> There's only really ever one Theme/Page/Style in use - the live one. The dev state is just a convenience for iterating a design, so it's probably not worth the upheaval to change the labels at the moment.

That makes sense now after yet more exploration.

> Oh, one more thing, where did mention of the Deploy to live / Reset to live multi-edit option go? It only appears if you have set a dev theme.

There are some adjusted/new sections under the [Production (live) themes](https://docs.textpattern.com/build/themes-creating-using-and-sharing#production-live-themes) section. See if that's what you're referring to. The information was there before, but just not called out plainly by a clear header.

What I'm recognizing is there is no real way to theme switch all in one shot unless a site is setup to use a single page and style across every section. Otherwise, a person will have to make multiple passes with the assigning controls, regardless whether a single theme across the whole site or a different theme on every section.

By the end of the 'Production (live) themes' sections, after having already read through the development sections before it, the explanations of using the selection controls border on being beat to death; only different by the context of single theme or multiple themes. But I think the consolidation is as much as can be while still remaining clear under the different contexts or objectives people might have.

I need to check internal links again, but I'll do that tomorrow.

**Comment by Bloke** [comment]
Yeah, sort of. But that's where the Deploy to Live multi-edit option comes in. It allows you to, essentially, _stage_ your entire setup and then switch it live in one go.

So you can, for example:

* set Sections A, B, D to use _dev_ Theme 'downloaded-theme'.
* set Sections C and E to use _dev_ Theme 'my-new-theme'.

Preview everything using your logged-in powers, make sure everything hangs together, then from the Sections panel:

* Select Sections A thru E.
* Choose 'Developer preview' from the multi-edit list.
* Select 'Deploy to Live'.
* Go + confirm.

All your selected sections will get the current _dev_ theme copied to _live_. One shot.

If at some point during development, you or your client decide that the direction things are heading is a load of crap and you want to start again on a bunch of sections, select the sections you want to "unstage", choose 'Developer preview' from the multi-edit list, and select 'Reset to Live'. The current Dev theme(s) will be unassigned, thus the previous Live theme(s) will be shown to logged-in users.

Yes, you still have to set things up and do multiple passes to assign multiple themes as you develop the site, but the critical thing is that you don't have to make all the changes quickly and leave your site in potential limbo at go-live time. The multi-edit option will copy dev -> live - regardless of which theme/page/style you've used - in one go for the selected sections.

Does that make sense?

**Comment by wion** [comment]
> * Choose 'Developer preview' from the multi-edit list.

I don't see any such option in either 4.8.0-dev demo or my local 4.8.0-dev installation. It must really be bleeding edge.

**Comment by Bloke** [comment]
It only appears after you've made at least one theme dev.

**Comment by wion** [comment]
I must have overlooked it then, because I've made a lot of themes dev in the course of working on this document. Clearly it needs to be documented.

A suggestion for the future, if I may... That for any new, significant functionality to be introduced in the software and needing documented; that it be _briefly_ summarized in a development spec, or whatever, principally naming what new functionality it includes and _where_ to find it and _when_ to see it in the peek-a-boo nature of hiding things in the UI. 

I suspect this themes project was one of the more challenging that Txp will ever see, and once done there won't be a greater effort, but if, say, unlimited custom fields is going to be another bear to document, it would be nice to have such a spec first for the tech writer, to be more efficient than foolish.

I will investigate and edit further.

**Comment by wion** [comment]
Oh, that looks kind of familiar. ;)

But now I'm confused why there is even a 'Live' check box option in the 'Change theme/page/style' controls. I thought that was the new way of deploying a theme asset to live on a section.

**Comment by wion** [comment]
Anyway, I think I just need to add a new 'Deploy to live' section in the Production part of the doc and that should be good enough.

**Comment by Bloke** [comment]
> Oh, that looks kind of familiar. ;)
> 
> But now I'm confused why there is even a 'Live' check box option in the 'Change theme/page/style' controls. I thought that was the new way of deploying a theme asset to live on a section.

It's _a_ way of deploying stuff live. If you only have one section to push live, or if you have a few sections that share a single theme, then you can just do it there. Choice of pushing a theme to live/dev is handy.

But if you want to stage the entire site with multiple themes across many sections, the Developer preview multi-edit offers the least downtime when switching over from dev->live. Just one action.

The Developer preview info was in the doc in an earlier iteration (at least I thought it was) so maybe it got chopped out during an edit storm. Or, perhaps during testing as the docs was being developed, the option 'disappeared' because it only shows up if there are development themes assigned. We need to do this, because the option doesn't make sense otherwise.

**Comment by wion** [comment]
> The Developer preview info was in the doc in an earlier iteration (at least I thought it was) so maybe it got chopped out during an edit storm.

That's exactly what happened. And I chopped it under the confused impression that functionality was changing, then I just overlooked it every time. But I'm glad the preview functionality is there and understand it better; it cleared yet more things up for me.

And despite the loopy-loop race-track editing, I think we're close now, and it's even better than ever. The document is a bit more concise in places, progresses more logically, has clearer section links, for the most part, with more intra-document linking.

I need to remove some obsolete images now, add a couple of new/improved ones, and make another copyedit and link-check pass, then I'll ping you for one more golden-seal review before seeking the public impression.

**Comment by wion** [comment]
4.8.0 in the house! Good job, everyone.

I’ll get back to this doc as soon as the kids return to school on Monday.
